src/libutil/args.hh

```diff
@@ -278,21 +341,37 @@ struct Completion {
 
     bool operator<(const Completion & other) const;
 };
-class Completions : public std::set<Completion> {
+
+/**
+ * The abstract interface for completions callbacks
+ *
+ * They can only add additional completions to the collection, or set
+ * the completion type.
+ */
+class AddCompletions
```

Comment:
We don't need this, but this is what I was thinking:

If we give the callback access to the concrete `Completions` type, then it could change that data in arbitrary ways. The `AddCompletions` interface forces it to just do these two operations, so it can't mess with any other compleitions. Then only the code that needs to process all the completions gets access to the underlying set in order to do so.